crates/ruff/src/jupyter/notebook.rs

```diff
@@ -130,60 +168,137 @@ impl JupyterNotebook {
+        for edit in edits.into_iter().rev() {
+            let idx = self
+                .cell_offsets
+                .iter()
+                .tuple_windows::<(_, _)>()
+                .find_position(|(&offset1, &offset2)| {
+                    offset1 <= edit.start() && edit.end() <= offset2
+                })
+                .map_or_else(
+                    || panic!("edit outside of any cells: {edit:?}"),
+                    |(idx, _)| idx,
+                );
+
+            if edit.is_deletion() {
+                for offset in &mut self.cell_offsets[idx + 1..] {
+                    *offset -= edit.range().len();
                 }
-                SourceValue::StringArray(string_array) => {
-                    jupyter_index.row_to_cell.extend(
-                        iter::repeat(u32::try_from(pos + 1).unwrap()).take(string_array.len()),
-                    );
-                    jupyter_index
-                        .row_to_row_in_cell
-                        .extend(1..=u32::try_from(string_array.len()).unwrap());
-                    // lines already end in a newline character
-                    string_array.join("")
+            } else if edit.is_insertion() {
+                let new_text_size = TextSize::of(edit.content().unwrap_or_default());
+                for offset in &mut self.cell_offsets[idx + 1..] {
+                    *offset += new_text_size;
                 }
-            };
+            } else if edit.is_replacement() {
+                let current_text_size = edit.range().len();
+                let new_text_size = TextSize::of(edit.content().unwrap_or_default());
+                match new_text_size.cmp(&current_text_size) {
+                    Ordering::Less => {
+                        for offset in &mut self.cell_offsets[idx + 1..] {
+                            *offset -= current_text_size - new_text_size;
+                        }
+                    }
+                    Ordering::Greater => {
+                        for offset in &mut self.cell_offsets[idx + 1..] {
+                            *offset += new_text_size - current_text_size;
+                        }
+                    }
+                    Ordering::Equal => (),
+                };
+            } else {
+                panic!("unexpected edit: {edit:?}");
+            }
+        }
+    }
+
+    fn update_cell_content(&mut self, transformed: &str) {
+        for (&pos, (start, end)) in self
+            .valid_code_cells
+            .iter()
+            .zip(self.cell_offsets.iter().tuple_windows::<(_, _)>())
+        {
+            let cell_content = transformed
+                .get(start.to_usize()..end.to_usize())
+                .unwrap_or_else(|| panic!("cell content out of bounds: {:?}", &self.raw.cells[pos]))
+                .trim_end_matches(|c| c == '\r' || c == '\n')
+                .to_string();
+            self.raw.cells[pos].source = SourceValue::String(cell_content);
```

Comment:
The reason to choose `String` over `StringArray` is that the latter should contain the newline as part of each element which means that `str.lines` would not work and we would either need a custom iterator or add newlines to each element individually.